book/src/generalized-ownership.md

```diff
@@ -0,0 +1,279 @@
+# Generalized Ownership
+
+This design describes generalized ownership of countable items on the Solana
+blockchain. The goal of this design is to propose a single general-purpose
+program whose state can be used across many different programs to implement
+ownership, transfer and programmatic assignment. 
+
+Many programs in Solana should be able to use a common program similar to
+`System::Move`.
+
+## Current System and Token programs
+
+System program defines ownership of lamports to be identified by the public key
+that matches the account's pubkey. The`System::Move`instruction checks that
+the identity of the spending account has a valid signature before completing the
+move.
+
+Token program defines ownership of the token to be identified by the public key
+that matches the account's pubkey, and`Token::Transfer`instruction checks
+that the identity of the spending account has a valid signature before
+completing the move.
+
+Both of these programs implement almost the same mechanism.
+
+## Goal for this design
+
+The goal of this design is to make the following possible:
+
+In a single program, it should be possible to change and update ownership of
+arbitrary units of tokens implemented by other programs.
+
+Alice requests X amount of token implemented by program Token1, in
+exchange for Y amount of token implemented by program Token2.
+
+Bob offers X + 1 amount of token implemented by program Token1, in exchange for
+Y - 1 amount of token implemented by Token2.
+
+Dan pairs Bob and Alice, and the result ownership changes:
+
+* Alice owns X of Token1,
+* Bob owns Y - 1 of Token2,
+* Dan owns 1 of Token1 and 1 of Token2,
+
+Another example:
+
+Alice offers X amount of Token1 to the winner of TicTacToe.
+
+Bob offers Y amount of Token2 to the winner of the same game of TicTacToe.
+
+The game charges a commission, the ceiling of the percentage of the tokens.  The
+winner takes ownership of both tokens minus the commission.  The commission is
+transferred to the address designated by the game.
+
+## General idea
+
+For items to have externally countable ownership, they must rely on the external
+program for managing the transfer of units.  These items must have a valid state
+where the number of units owned is 0.  The **zero** count of items is
+the destination for transferring any count of items on the chain.
+
+Execution of these state transitions must allow for a re-entrant and partial
+transactions to make progress.
+
+## Generalized owner identity
+
+Using the account's pubkey identity as the asset owner identity means that all
+the asset transfer APIs must be re-implemented by the program code.
+
+The proposed solution is to use a pointer to an owner identity.
+
+```rust,ignore
+// This state is owned by the Owner program
+enum OwnerState {
+    // the owner identity must provide a valid signature for this pubkey
+    Unique {
+        // The actual owner key.
```

Comment:
"actual owner" is a strange thing to say. Smells of a terminology problem. Is there a "not-actual owner" that's hijacked the term "owner"?

Reply:
Account has an owner, which is the program.  Signer, which is the thing that owns the some keypair.  The owner of the state which stores these bits, which is irrelevant here.  How about `token_owner`, or `unit_owner`?

Reply:
I can't help specialize vague terminology. I asked because I can't tell what you're referring to. I need you to spend some time either thinking on the terminology and defining it earlier in this document, or spend more time clarifying the problem statement and have others propose solutions.